Sources/TransactionalProperty.swift

```diff
@@ -0,0 +1,690 @@
+		_ body: @escaping (Value) -> Result<T.Value, TransactionError>
+	) where T.TransactionError == NoError {
+		self.init(inner, transform: transform, validator: body, validationSetup: nil)
+	}
+
+	private init<T: _TransactionalPropertyProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		validator: @escaping (Value) -> Result<T.Value, TransactionError>,
+		validationSetup: ((MutableProperty<ValidationResult<Value, TransactionError>>) -> Void)?
+	) {
+		let _validations: MutableProperty<ValidationResult<Value, TransactionError>> = inner.withValue { innerValue in
+			let property = MutableProperty(ValidationResult(innerValue, transform: transform, validator: validator))
+			validationSetup?(property)
+			return property
+		}
+
+		self.validations = Property(capturing: _validations)
+		self.lifetime = inner.lifetime
+		self.cache = inner.property.cache.map(transform)
+		self.rootBox = inner.property.rootBox
+
+
+		action = { input in
+			switch validator(input) {
+			case let .success(innerResult):
+				inner.property.action(innerResult)
+
+			case let .failure(error):
+				_validations.value = .failure(input, error)
+			}
+		}
+	}
+
+	/// Atomically performs an arbitrary action using the current value of the
+	/// variable.
+	///
+	/// - parameters:
+	///   - action: A closure that accepts current property value.
+	///
+	/// - returns: the result of the action.
+	public func withValue<Result>(action: (Value) throws -> Result) rethrows -> Result {
+		return try rootBox.lock {
+			return try action(cache.value)
+		}
+	}
+
+	/// Atomically modifies the variable.
+	///
+	/// - parameters:
+	///   - action: A closure that takes the current value.
+	///
+	/// - returns: The result of the action.
+	public func modify<Result>(_ action: (inout Value) throws -> Result) rethrows -> Result {
+		return try rootBox.lock {
+			return try action(&value)
+		}
+	}
+
+	internal func revalidate() {
+		rootBox.lock {
+			action(cache.value)
+		}
+	}
+}
+
+public protocol _TransactionalPropertyProtocol: ComposableMutablePropertyProtocol {
+	associatedtype Value
+	associatedtype TransactionError: Swift.Error
+
+	var property: TransactionalProperty<Value, TransactionError> { get }
+
+	init<T: _TransactionalPropertyProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		errorTransform: @escaping (T.TransactionError) -> TransactionError,
+		_ body: @escaping (Value) -> Result<T.Value, TransactionError>
+	)
+
+	init<T: _TransactionalPropertyProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		_ body: @escaping (Value) -> Result<T.Value, TransactionError>
+	) where T.TransactionError == NoError
+}
+
+extension TransactionalProperty: _TransactionalPropertyProtocol {
+	public var property: TransactionalProperty<Value, TransactionError> {
+		return self
+	}
+}
+
+public enum ValidationResult<Value, Error: Swift.Error> {
+	case success(Value)
+	case failure(Value, Error)
+
+	public var isFailed: Bool {
```

Comment:
`didFail` would be a better name for this, or `isFailure` if you really want to match the case label.